compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -119,6 +119,46 @@ mod sty;
 
 // Data types
 
+/// A pointer to a value that is known to be unique, e.g. via interning.
+/// The uniqueness property means that equality and hashing can be done on the
+/// pointer rather than the value's contents, which can improve performance.
+#[derive(Clone, Copy, Debug, HashStable)]
+pub struct Uniq<'tcx, T>(&'tcx T);
```

Comment:
Can you move this to a submodule, say `ty/uniq.rs`? This is fairly independent of the rest of the module, and its implementation should be hidden from the rest of `rustc_middle::ty` IMO. Also, this file is already >2000 LOC.

Reply:
Fair enough, I will do this.

Reply:
BTW, I see in `rustc_data_structures` there is already a `PtrKey` type that is very similar. I'll look into its uses and see if `Uniq` should replace that.